packages/engine/crawlers/solana/transactionHistory.ts

```diff
@@ -0,0 +1,515 @@
+		(item) => item.owner !== ownerPublicKey,
+	);
+
+	if (type === 'sent') {
+		sender = ownerPublicKey;
+		receiver = partner[0].owner ?? '';
+	} else {
+		sender = partner[0].owner ?? '';
+		receiver = ownerPublicKey;
+	}
+
+	return {
+		sender,
+		receiver,
+	};
+};
+
+export const getTransactionDetails = async (
+	solanaContext: SolanaContext,
+	parsedTransaction: ParsedTransactionWithMeta,
+	ownerPublicKey: string,
+) => {
+	if (!parsedTransaction.meta) return;
+
+	const instructions = parsedTransaction.transaction.message.instructions;
+	const paymentInstruction = instructions[0];
+	const mainInstruction = instructions[instructions.length - 1];
+
+	const confirmation = await getConfirmation(
+		solanaContext,
+		parsedTransaction.transaction.signatures[0],
+	);
+
+	const status =
+		confirmation === 'finalized' || confirmation === 'confirmed'
+			? 'success'
+			: confirmation === 'processed'
+			? 'pending'
+			: 'failed';
+
+	const type = getTransactionType(parsedTransaction, ownerPublicKey);
+
+	const { sender, receiver } = getTransactionPartners(
+		parsedTransaction,
+		ownerPublicKey,
+		type,
+	);
+
+	const token = await getTransactionToken(
+		mainInstruction,
+		parsedTransaction.meta,
+		type,
+		sender,
+		receiver,
+		solanaContext,
+	);
+
+	const tokenForFee = await getTransactionTokenForFee(
+		paymentInstruction,
+		parsedTransaction.meta,
+		solanaContext,
+	);
+
+	const fee = getTransactionFee(parsedTransaction, tokenForFee);
+
+	const { preBalance, postBalance } = getTransactionBalances(
+		parsedTransaction.meta,
+		ownerPublicKey,
+	);
+
+	const amount = getTransactionAmount({
+		preBalance,
+		postBalance,
+	});
+
+	const finalTransaction: Transaction = {
+		id: parsedTransaction.transaction.signatures[0],
+		signature: parsedTransaction.transaction.signatures[0],
+		network: Networks.solana,
+		type,
+		status,
+		sender,
+		receiver,
+		token,
+		tokenForFee,
+		fee: fee ?? 0,
+		preBalance,
+		postBalance,
+		amount,
+		date: parsedTransaction.blockTime
+			? new Date(parsedTransaction.blockTime * 1000)
+			: new Date(),
+	};
+
+	return finalTransaction;
+};
+
+export const getSignatureList = async (
+	solanaContext: SolanaContext,
+	ownerPublicKey: string,
+) => {
+	const accounts = await solanaContext.connection.getParsedProgramAccounts(
+		TOKEN_PROGRAM_ID,
+		{
+			filters: [
+				{
+					dataSize: 165,
+				},
+				{
+					memcmp: {
+						offset: 32,
+						bytes: ownerPublicKey,
+					},
+				},
+			],
+		},
+	);
+
+	const tokenAccounts = [ownerPublicKey];
+	accounts.map((account) => {
+		tokenAccounts.push(account.pubkey.toString());
+	});
+
+	let confirmedSignatureInfos: ConfirmedSignatureInfo[] = [];
+	for (let i = 0; i < tokenAccounts.length; i++) {
+		const signatureInfos =
+			await solanaContext.connection.getSignaturesForAddress(
+				new PublicKey(tokenAccounts[i]),
+				{ limit: 30 },
+			);
+		confirmedSignatureInfos = Array.from(
+			new Set(confirmedSignatureInfos.concat(signatureInfos)),
+		);
+	}
```

Comment:
could we use `forEach` here?

Reply:
i need the index so i use this kind of iteration